docs/contributor/writing-effective-tests.md

```diff
@@ -0,0 +1,316 @@
+# Writing Effective Tests <!-- omit in toc -->
+
+This document provides guidance to contributors on how to create effective test cases for contributions to the [VSCode Swift extension](https://github.com/swiftlang/vscode-swift) using **Mocha**, **Chai**, and **Sinon**. These tools are widely used for testing JavaScript/TypeScript code and will help ensure that your code is reliable and maintainable.
+
+A brief description of each framework can be found below:
+
+-   [**Mocha**](https://mochajs.org/): A JavaScript test framework that runs in Node.js. It provides structure for writing test suites, hooks (e.g., `setup()`, `teardown()`), and test cases (`test()`).
+-   [**Chai**](https://www.chaijs.com/): An assertion library. It allows you to express expectations for your code’s behavior using natural language.
+-   [**Sinon**](https://sinonjs.org/): A powerful mocking and spying library that allows you to create spies, stubs, and mocks to test how your code interacts with other components.
+
+## Overview <!-- omit in toc -->
+
+- [Organizing Tests](#organizing-tests)
+- [Running the Tests](#running-the-tests)
+- [Writing Unit Tests](#writing-unit-tests)
+- [Mocking Utilities](#mocking-utilities)
+  - [Mocking interfaces, classes, and functions](#mocking-interfaces-classes-and-functions)
+  - [Mocking VS Code events](#mocking-vs-code-events)
+  - [Mocking modules](#mocking-modules)
+    - [Mocking global interfaces](#mocking-global-interfaces)
+    - [Mocking global events](#mocking-global-events)
+    - [Setting global constants](#setting-global-constants)
+    - [Mocking an entire module](#mocking-an-entire-module)
+- [Conclusion](#conclusion)
+
+## Organizing Tests
+
+Tests are currently organized into one of two directories:
+
+1. **Unit Tests** - Tests that exercise smaller units of code go in `test/unit-tests`
+2. **Integration Tests** - Tests that exercise more complicated features that require the extension to be loaded in VSCode go in `test/integration-tests`
+
+There are plans to add a third set of smoke tests that will have no mocking at all and actually click through the VS Code user interface to perform actions. For more information see the [Swift for Visual Studio Code test strategy](./test-strategy.md) document.
+
+## Running the Tests
```

Comment:
If the point of this guide is to highlight how to run the tests, maybe move this to the contributors guide

Reply:
Yeah that makes sense. I reworded the `CONTRIBUTING.md` document to reference this document.